platform/api.scripting/src/org/netbeans/api/scripting/Scripting.java

```diff
@@ -0,0 +1,457 @@
+ * returns the desired Java type directly, as shown in the following variation.
+ * <p>
+ * {@codesnippet org.netbeans.api.scripting.ScriptingTutorial#accessFieldsOfJavaObjectWithConverter}
+ *
+ * <h3>Access Java constructors and static methods from JavaScript</h3>
+ *
+ * Dynamic languages can also access the <em>public constructors</em> and <em>public static methods</em>
+ * of any Java class for which they are provided a reference.
+ * The following example shows JavaScript access to the public constructor of a Java
+ * class.
+ * <p>
+ * {@codesnippet org.netbeans.api.scripting.ScriptingTutorial#createJavaScriptFactoryForJavaClass}
+ *
+ * Notes:
+ * <ul>
+ * <li>Evaluating the JS source returns an anonymous JS function of one argument
+ * assigned to {@code jsFunction} that can be executed directly with one argument.</li>
+ * <li>The Java class argument {@code Moment.class} is seen by the JS function as a
+ * <em>foreign class</em> whose public constructor is visible.</em>
+ * <li>Executing {@code jsFunction} with the Java class argument returns
+ * a JS "factory" function (for the Java class) that can be
+ * {@linkplain javax.script.Invocable#getInterface "cast"}
+ * to the desired Java function type ({@code MomentFactory}).</li>
+ * <li>Parentheses around the JS function definition keep it out of JavaScript's
+ * global scope, so the Java object holds the only reference to it.</li>
+ * </ul>
+ *
+ * </div>
+ * <script src="doc-files/tutorial.js"></script>
+ *
+ * @since 1.0
+ */
+public final class Scripting {
+    private Scripting() {
+    }
+
+    /** Create new {@link ScriptEngineManager} configured for the NetBeans
+     * environment. The manager serves as an <em>isolated</em> environment -
+     * engines created from the same manager are supposed to share the
+     * same internals and be able to communicate with each other.
+     *
+     * @return new instance of the engine manager
+     */
+    public static ScriptEngineManager createManager() {
+        List<ScriptEngineFactory> extra = new ArrayList<>();
+        for (EngineProvider p : Lookup.getDefault().lookupAll(EngineProvider.class)) {
+            extra.addAll(p.factories());
+        }
+        ClassLoader l = Lookup.getDefault().lookup(ClassLoader.class);
+        if (l == null) {
+            l = Thread.currentThread().getContextClassLoader();
+        }
+        if (l == null) {
+            l = Scripting.class.getClassLoader();
+        }
+        return new EngineManager(extra, l);
+    }
+
+    private static final class EngineManager extends ScriptEngineManager {
+        private final List<ScriptEngineFactory> extra;
+
+        EngineManager(List<ScriptEngineFactory> extra, ClassLoader loader) {
+            super(loader);
+            this.extra = extra;
+
+            for (ScriptEngineFactory f : extra) {
+                registerEngineName(f.getEngineName(), f);
+                for (String ext : f.getExtensions()) {
```

Comment:
Original `ScriptEngineManager` impl (APIspec is silent) permits `null` outcome of `getExtensions` and `getMimeTypes`, could be more foolproof here as well.

Reply:
Let's say that extra engines that are provided by `EngineProvider` have to return non-null.